### PR TITLE
Fix sampling strategies overwriting service entry when no sampling ty…

### LINKED ISSUE #1205: Multiple service sampling strategies broken
When we set multiple services in `service_strategies`, the last one is shown in the response returned by agent for every service queried.

For example, when the sampling strategies file is this: 
```
"default_strategy": {"param": 0.10000000000000001, "type": "ratelimiting"},
    "service_strategies": [{"operation_strategies": [{"operation": "HTTP GET", "param":
    1, "type": "probabilistic"}, {"operation": "HTTP GET - metrics", "param": 0, "type":
    "probabilistic"}], "service": "query-frontend"}, {"operation_strategies": [{"operation":
    "HTTP GET - metrics", "param": 1.0000000000000001e-05, "type": "probabilistic"}],
    "service": "ingester"}]}
```

The response from agent:
```
# wget -qO- http://jaeger-agent.jaeger.svc:5778/sampling?service=query-frontend
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":0.001},"operationSampling":{"defaultSamplingProbability":0.001,"defaultLowerBoundTracesPerSecond":0,"perOperationStrategies":[{"operation":"HTTP GET - metrics","probabilisticSampling":{"samplingRate":0.00001}}]}}
```

When the file is the following:
```
{"default_strategy": {"param": 0.10000000000000001, "type": "ratelimiting"},
    "service_strategies": [{"operation_strategies": [{"operation": "HTTP GET", "param":
    1, "type": "probabilistic"}, {"operation": "HTTP GET - metrics", "param": 0, "type":
    "probabilistic"}], "service": "query-frontend"}, {"operation_strategies": [{"operation":
    "HTTP GET - metrics", "param": 0, "type": "probabilistic"}], "service": "querier"},
    {"operation_strategies": [{"operation": "HTTP GET - metrics", "param": 1.0000000000000001e-05,
    "type": "probabilistic"}], "service": "ingester"}]}
```

The response from agent:
```
# wget -qO- http://jaeger-agent.jaeger.svc:5778/sampling?service=query-frontend
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":0.001},"operationSampling":{"defaultSamplingProbability":0.001,"defaultLowerBoundTracesPerSecond":0,"perOperationStrategies":[{"operation":"HTTP GET - metrics","probabilisticSampling":{"samplingRate":0.00001}}]}}

# wget -qO- http://jaeger-agent.jaeger.svc:5778/sampling?service=querier
{"strategyType":"PROBABILISTIC","probabilisticSampling":{"samplingRate":0.001},"operationSampling":{"defaultSamplingProbability":0.001,"defaultLowerBoundTracesPerSecond":0,"perOperationStrategies":[{"operation":"HTTP GET - metrics","probabilisticSampling":{"samplingRate":0.00001}}]}}
```

This is on `v1.8.0` of jaeger.

### PR DESCRIPTION
**Pull request by objectiser**
…pe is specified

Signed-off-by: Gary Brown <gary@brownuk.com>

## Which problem is this PR solving?
Resolves #1205 
If the service strategy does not define a sampling type, it reuses the default sampling strategy but applies the operation sampling strategies. End result is that all services (that don't define a sampling type) will share a single entry with identical details (i.e. the ones associated with the last initialised entry).

## Short description of the changes
Although it is an error state, if no sampling type is defined and the default sampling strategy is to be used, then it should return a copy.

### CODE REVIEW

**Comment by yurishkuro** [comment]
> If the service strategy does not define a sampling type,

what do you mean by "sampling type"?

Taking a copy seems fine, but what does it fix? Do we modify the returned struct somehow, so that the default is affected for every service?

**Comment by objectiser** [comment]
For example, as shown [here](https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/static/fixtures/operation_strategies.json#L9). The service level sampling type.

Yes the [defaultStrategy](https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/static/fixtures/operation_strategies.json#L9) is being modified.

**Comment by objectiser** [comment]
Maybe not in this PR, but I think in general it would be better to enable the service specific sampling strategy to be optional, and if not defined, then the [default defined in the config](https://github.com/jaegertracing/jaeger/blob/master/plugin/sampling/strategystore/static/fixtures/operation_strategies.json#L2-L5) should be used. Only if that is not defined, should the hardcoded default be used.

File: plugin/sampling/strategystore/static/strategy_store.go [hunk 1]
```diff
@@ -136,6 +136,7 @@ func (h *strategyStore) parseStrategy(strategy *strategy) *sampling.SamplingStra
 		}
 	default:
 		h.logger.Warn("Failed to parse sampling strategy", zap.Any("strategy", strategy))
-		return &defaultStrategy
+		copyOfDefaultStrategy := defaultStrategy
+		return &copyOfDefaultStrategy
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/static/strategy_store.go]: I am not sure this way of copying is correct. It only works for primitive strategies, anything nested would still have the same issue. I think it needs a deep copy.

**Comment by objectiser** [comment]
@yurishkuro What is the preferred approach for doing deep copy in Jaeger?

**Comment by yurishkuro** [comment]
We haven't done it anywhere afaik. One option is to use `encoding/glob` to marshal/unmarshal.

**Comment by objectiser** [comment]
@yurishkuro Updated.

**Pull request merged**